### ISSUE TITLE
Support LLaMA Models, including HuggingFace-adapted variants

### ISSUE THREAD

**Issue by zoidbb** [BlackSamorez/tensor_parallel#51]
Facebook has released a new model known as LLaMA. It would be awesome to use tensor_parallel with it. Currently getting the following error when trying to use decapoda-research/llama-30b-hf:

``
RuntimeError: The size of tensor a (1110) must match the size of tensor b (6656) at non-singleton dimension 2
```

**Comment by BlackSamorez** [comment]
@zoidbb  I need more information on how to reproduce this error to be able to solve it.
Since there is no predefined config for LLaMA models the auto generated config is to be used and it should always work fine.
Were you trying to use [LLaMA code that is being added to transformers](https://github.com/huggingface/transformers/pull/21955)?

**Comment by BlackSamorez** [comment]
I have successfully ran `LLaMa 7B` with autoconfig on `2xT4` in a [kaggle notebook](https://www.kaggle.com/code/blacksamorez/llama-7b) without any model-specific code.
It's obviously suboptimal but I'll probably create a custom config for those models some time later.

**Comment by zoidbb** [comment]
Unfortunately Kaggle says that notebook doesn't exist.

I'm using a machine with 6sxA6000 cards.

```
from transformers import AutoModelForCausalLM
import tensor_parallel as tp

model = AutoModelForCausalLM.from_pretrained("decapoda-research/llama-7b-hf" device_map='auto'
model = tp.tensor_parallel(model, [f'cuda:{i}' for in range(6)])
model.generate(some_input_ids) # this is where it breaks.
```

This is basically what I'm doing, using the transformers [PR that supports LLaMA](https://github.com/huggingface/transformers/pull/21955)

**Comment by BlackSamorez** [comment]
> Unfortunately Kaggle says that notebook doesn't exist.
> 
> I'm using a machine with 6sxA6000 cards.
> 
> ```
> from transformers import AutoModelForCausalLM
> import tensor_parallel as tp
> 
> model = AutoModelForCausalLM.from_pretrained("decapoda-research/llama-7b-hf" device_map='auto'
> model = tp.tensor_parallel(model, [f'cuda:{i}' for in range(6)])
> model.generate(some_input_ids) # this is where it breaks.
> ```
> 
> This is basically what I'm doing, using the transformers [PR that supports LLaMA](https://github.com/huggingface/transformers/pull/21955)

I've fixed the notebook visibility issue. It should be public now.
By the way, I'm using the exact same model and the exact same transformers code for it and it works on 2 GPUs and 4 GPUs. Maybe it's something that only occurs on 6 GPUs.

**Comment by UranusSeven** [comment]
I'm able to run LLaMA-7B with minor changes on the [notebook](https://www.kaggle.com/code/blacksamorez/llama-7b).

Here's my code:
```python
# torch version 2.0.0
import torch
# tensor-parallel version 1.0.22
from tensor_parallel import TensorParallelPreTrainedModel
# transformer version 4.28.0.dev0
from transformers import LlamaTokenizer, LlamaForCausalLM, GenerationConfig

model = LlamaForCausalLM.from_pretrained("decapoda-research/llama-7b-hf", torch_dtype=torch.float16)
model = TensorParallelPreTrainedModel(model, ["cuda:0", "cuda:1"])

# I got some warnings:
# The tokenizer class you load from this checkpoint is not the same type as the class this function is called from. It may result in unexpected tokenization. The tokenizer class you load from this checkpoint is 'LLaMATokenizer'. The class this function is called from is 'LlamaTokenizer'.
tokenizer = LlamaTokenizer.from_pretrained("decapoda-research/llama-7b-hf")

tokens = tokenizer("Hi, how are you?", return_tensors="pt")
tokenizer.decode(model.generate(tokens["input_ids"].cuda(0), attention_mask=tokens["attention_mask"].cuda(0))[0])
```

Seems tensor-parallel blances the varm pretty well:
![image](https://user-images.githubusercontent.com/109661872/229843024-2b130aac-0e95-4a38-85de-2cf2d5b51d2e.png)

**Comment by BlackSamorez** [comment]
Yeah, `LLaMA` is fully functional and `llama` branch provides the optimal config.
Here's the [demo with LLaMA 13B on kaggle](https://www.kaggle.com/code/blacksamorez/tensor-parallel-llama-13b/).
I'm waiting for a proper `transformer version 4.28.0` release to merge `llama` into `main`.
